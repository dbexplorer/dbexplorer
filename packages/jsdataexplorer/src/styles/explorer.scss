.jsde-explorer {
    font-family: sans-serif;
    display: flex;
    gap: 16px;
    height: 100%;
}
.jsde-panel {
    border: 1px solid;
    position: relative;
    overflow-x: clip;
    height: 100%;
    display: flex;
    flex-direction: column;
    width: min-content;
}
.jsde-panel-body {
    overflow-y: auto;
    flex-grow: 1;
    height: 0;
    display: flex;
}
.jsde-panel__close {
    position: absolute;
    margin: 4px;
    width: 24px;
    height: 24px;
    right: 0;
    top: 0;
    padding: 0;
    display:flex;
    justify-content: center;
    align-items: center;
    &::after {
        content: '\2715';
    }
    z-index: 1;
}
.jsde-panel__header {
    margin: 4px;
    height: 24px;
    font-size: 24px;
    position: sticky;
    top: 0;
}