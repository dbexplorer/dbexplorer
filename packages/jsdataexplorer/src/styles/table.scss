.jsde-table {
    height: 0;
}

.jsde-table-head {
    color: white;
    background-color: black;
    position: sticky;
    top: 0;
}

.jsde-table-row {
    &:hover {
        background-color: gray;
    }
}

.jsde-table-row--explored {
    background-color: lightgrey;
}

.jsde-table-cell {
    padding: 8px;
    white-space: nowrap;
}