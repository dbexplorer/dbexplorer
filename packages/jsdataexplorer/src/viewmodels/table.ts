import { IGetDataOptions } from "../data";
import { cssPrefix, keysEqual } from "../utils";

export class TableCellViewModel {
  private text: string;
  getText() {
    return this.text;
  }
  setText(text: string) {
    this.text = text;
    this.updateCallback(this.text);
  }
  constructor(text: string = null) {
    this.text = text;
  }
  css() {
    return { root: cssPrefix("table-cell") };
  }
  updateCallback: (data: string) => any = () => { };
}
export class TableRowViewModel {
  private cells: TableCellViewModel[];
  private explored: boolean;
  getKey() {
    return this.key as string;
  }
  getCells() {
    return this.cells;
  }
  setCellsText(data: string[]) {
    if (!this.cells) {
      this.cells = data.map((s) => new TableCellViewModel(s));
    }
    else {
      data.map((s, idx) => this.cells[idx].setText(s));
    }
  }
  setExplored(explored: boolean, needUpdate: boolean = true) {
    this.explored = explored;
    if (needUpdate) this.updateCssCallback(this.css().root);
  }

  cssModifiers() {
    return {
      root: {
        explored: "--explored"
      }
    };
  }
  css() {
    return { root: cssPrefix("table-row") + (this.explored ? " " + cssPrefix("table-row") + this.cssModifiers().root.explored : "") };
  }
  constructor(data: string[], private key: string | string[]) {
    this.setCellsText(data);
  }
  updateDataCallback: (data: TableCellViewModel[]) => any;
  updateCssCallback: (css: string) => any;
  exploreCallback: () => void;
}

export class TableHeaderViewModel {
  public captionsViewModel: TableRowViewModel;
  constructor(captions: string[]) {
    this.captionsViewModel = new TableRowViewModel(captions, null);
  }
}

export class TableColumnDescription {
  name: string;
  title?: string;
  hasReference?: boolean;
  isMain?: boolean;
}

export class TableViewModel {
  private columns: TableColumnDescription[];
  private dataOffset: number = 0;
  private dataBackOffset: number = 0;
  private exploredRow: TableRowViewModel = null;
  private loadBackVisible = true;
  dataPartRowCount: number = 10;
  headerViewModel: TableHeaderViewModel;
  private setLoadBackVisible(value: boolean) {
    this.loadBackVisible = value;
    this.loadBackVisibleCallback(false);
  }
  loadData(back = false) {
    const options = { limit: this.dataPartRowCount, offset: back ? this.dataBackOffset : this.dataOffset, key: this.key, back: back };

    if (back && !this.key) {
      this.setLoadBackVisible(false);
      return;
    }

    this.getDataCallback(
      options,
      (data: any) => {
        if (back) {
          this.dataBackOffset += this.dataPartRowCount;
        }
        else {
          this.dataOffset += this.dataPartRowCount;
        }
        if (data.length > 0) {
          const rows = data.map(
            (rowData: any) => {
              const row = new TableRowViewModel(
                this.columns.map((col) => rowData.data[col.name].toString()),
                rowData.key
              )
              if (keysEqual(rowData.key, this.key)) {
                this.exploredRow = row;
                this.exploredRow.setExplored(true, false);
                this.exploreRowCallback(row);
              }
              row.exploreCallback = () => {
                this.exploredRow?.setExplored(false);
                const newExploredRow = (this.exploredRow !== row) ? row : null;
                this.exploredRow = newExploredRow;
                this.exploredRow?.setExplored(true);
                this.exploreRowCallback(newExploredRow);
              }
              return row;
            }
          )
          if (back) {
            this.rows = rows.reverse().concat(this.rows);
          }
          else {
            this.rows = this.rows.concat(rows);
          }
          this.addRowsCallback(this.rows);
        }
        else {
          if (back) this.setLoadBackVisible(false); else this.loadMoreVisibleCallback(false);
        }
        if (!this.key) this.loadBackVisibleCallback(false);
      }
    );
  }
  rows: TableRowViewModel[];
  css() {
    return {
      root: cssPrefix("table"),
      head: cssPrefix("table-head"),
      body: cssPrefix("table-body"),
      foot: cssPrefix("table-foot"),
      loadingTop: cssPrefix("table-loading-top"),
      loadingBottom: cssPrefix("table-loading-bottom"),
    };
  }
  getDataCallback: (options: IGetDataOptions, ready: any) => void;
  constructor(columns: TableColumnDescription[], private key: any = undefined, private title = "Table") {
    this.columns = columns;
    this.headerViewModel = new TableHeaderViewModel(
      columns.map((col) => col.title || col.name)
    );
    this.rows = [];
  }
  getKey() {
    return this.key;
  }
  getKeyAsString() {
    return this.key.toString();
  }
  getTitle() {
    return this.title;
  }
  getCurrentRowIndex() {
    let index = this.rows.indexOf(this.rows.filter(r => keysEqual(r.getKey(), this.key))[0]);
    if (index >= 0 && this.loadBackVisible) index++;
    return index;
  }
  getLoadBackVisible(): any {
    return true;
  }
  getLoadMoreVisible(): any {
    return true;
  }
  addRowsCallback: (rows: TableRowViewModel[]) => any;
  loadMoreVisibleCallback: (visible: boolean) => any;
  loadBackVisibleCallback: (visible: boolean) => any = () => { };
  exploreRowCallback: (row: TableRowViewModel) => any;
}
