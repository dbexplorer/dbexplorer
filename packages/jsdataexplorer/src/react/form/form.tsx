import React, { useEffect } from 'react';
import { FormViewModel } from '../../viewmodels/form';
import { FormInputField } from './field';
import { FormRelationship } from './relationship';

export function Form({ model }: { model: FormViewModel }) {
  const css = model.css();
  useEffect(() => {
    model.reloadData();
  }, [])
  return (
    <form className={css.root} action="#">
      {model.fields.map((f, index) => <FormInputField key={index} model={f} />)}
      {model.rels.map((r, index) => <FormRelationship key={index} model={r} />)}
    </form>
  );
}