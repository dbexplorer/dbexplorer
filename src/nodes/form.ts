import { FormStringFieldViewModel, FormViewModel } from "../viewmodels/form";

export class FormInputFieldNode {
  private inputHTMLElement: HTMLInputElement;
  private labelHTMLElement: HTMLLabelElement;
  private fieldHTMLElement: HTMLDivElement;
  constructor(private viewModel: FormStringFieldViewModel) {
    this.labelHTMLElement = document.createElement("label");
    this.inputHTMLElement = document.createElement("input");
    this.fieldHTMLElement = document.createElement("div");
    this.fieldHTMLElement.appendChild(this.labelHTMLElement);
    this.fieldHTMLElement.appendChild(this.inputHTMLElement);

    this.labelHTMLElement.innerHTML = viewModel.getTitle();
    this.inputHTMLElement.innerHTML = viewModel.getText();
    viewModel.updateCallback = (data: string) => {
      this.inputHTMLElement.innerHTML = data;
    };
  }
  element(): HTMLElement {
    return this.fieldHTMLElement;
  }
}

export class FormNode {
  private formHTMLElement: HTMLFormElement;
  private fieldNodes: FormInputFieldNode[] = [];
  constructor(private viewModel: FormViewModel) {
    this.formHTMLElement = document.createElement("form");
    this.formHTMLElement.action = "#";
    viewModel.fields.forEach((field) => {
      let fieldNode = new FormInputFieldNode(field);
      this.fieldNodes.push();
      this.formHTMLElement.appendChild(fieldNode.element());
    });
  }
  element(): HTMLElement {
    return this.formHTMLElement;
  }
}
