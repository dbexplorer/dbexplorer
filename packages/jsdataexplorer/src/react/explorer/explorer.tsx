import React, { useState, useEffect } from 'react';
import { ExplorerViewModel } from '../../viewmodels/explorer';
import { ExplorerPanel } from '../explorer/panel';

export function Explorer({ model }: { model: ExplorerViewModel }) {
  const [panels, setPanels] = useState(model.getPanels());
  useEffect(() => {
    model.start();
  }, [])

  model.addPanelCallback = () => {
    setPanels([...model.getPanels()]);
  };
  model.removePanelCallback = () => {
    setPanels([...model.getPanels()]);
  };
  const css = model.css();
  return (
    <div className={css.root}>
      {panels.map((panel, index) => <ExplorerPanel key={panel.getKey()} model={panel} />)}
    </div>
  );
}