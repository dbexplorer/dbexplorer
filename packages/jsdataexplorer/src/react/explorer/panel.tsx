import React, { useState } from 'react';
import { IBaseViewModel } from '../../viewmodels/base';
import { ExplorerPanelViewModel } from '../../viewmodels/explorer';
import { FormViewModel } from '../../viewmodels/form';
import { TableViewModel } from '../../viewmodels/table';
import { Form } from '../form/form';
import { Table } from '../table/table';

export function ExplorerPanel({ model }: { model: ExplorerPanelViewModel }) {
  const [extraDataKey, setExtra] = useState("");

  model.setExtraDataKeyCallback = setExtra;
  const css = model.css();


  function handleCloseClick(e: any) {
    model.closeCallback();
  }

  return (
    <div className={css.root}>
      <button className={css.close} onClick={handleCloseClick}></button>
      <div className={css.header}>{model.dataViewModel.getTitle()}</div>
      <div className={css.body}>
        {<Table key={model.getKey()} model={model.dataViewModel as TableViewModel}></Table>}
        {extraDataKey ? <Form key={extraDataKey} model={model.extraDataViewModel as FormViewModel}></Form> : null}
      </div>
    </div>
  );
}