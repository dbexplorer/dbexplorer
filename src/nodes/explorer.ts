import { ExplorerPanelViewModel } from "../viewmodels/explorer";

import { BaseNode } from "../nodes/basenode";
import { TableNode } from "../nodes/table";
import { TableViewModel } from "../viewmodels/table";

export class ExplorerPanelNode {
  private panelHTMLElement: HTMLDivElement;
  private closeButtonHTMLElement: HTMLDivElement;
  private containerHTMLElement: HTMLDivElement;

  protected containerNode: BaseNode;

  protected createElements() {
    this.panelHTMLElement = document.createElement("div");
    this.closeButtonHTMLElement = document.createElement("div");
    this.containerHTMLElement = document.createElement("div");
    this.panelHTMLElement.appendChild(this.closeButtonHTMLElement);
    this.panelHTMLElement.appendChild(this.containerHTMLElement);
    this.containerHTMLElement.appendChild(this.containerNode.element());
  }

  constructor() {}
  element(): HTMLElement {
    return this.panelHTMLElement;
  }
}

export class ExplorerTablePanelNode extends ExplorerPanelNode {
  private tableNode: TableNode;
  constructor(private viewModel: ExplorerPanelViewModel) {
    super();
    this.containerNode = new TableNode(
      this.viewModel.dataViewModel as TableViewModel
    );
    this.createElements();
  }
}
