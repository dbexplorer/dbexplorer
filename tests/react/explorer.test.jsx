import React from 'react';
import { TableViewModel } from "../../src/viewmodels/table";
import { ExplorerPanelViewModel, ExplorerViewModel } from "../../src/viewmodels/explorer";
import { ExplorerPanel } from "../../src/react/explorer/panel"
import { Explorer } from "../../src/react/explorer/explorer"
import { FormViewModel } from "../../src/viewmodels/form";
import { render, fireEvent, waitFor, screen } from '@testing-library/react'
import { dbDescription } from "../schema_helper";
import { act } from 'react-dom/test-utils';
test("Table panel test", () => {
  var tableViewModel = new TableViewModel([]);
  var panelViewModel = new ExplorerPanelViewModel(tableViewModel);
  tableViewModel.getDataCallback = (limit, offset, ready) => {
    ready([]);
  };
  const { container } = render(<ExplorerPanel model={panelViewModel} />);
  expect(container.firstChild).toMatchSnapshot();
});

test("Form panel test", () => {
  var formViewModel = new FormViewModel([]);
  var panelViewModel = new ExplorerPanelViewModel(formViewModel);
  formViewModel.getDataCallback = (ready) => {
    ready([{}]);
  };
  const { container } = render(<ExplorerPanel model={panelViewModel} />);
  expect(container.firstChild).toMatchSnapshot();
});

test("Explorer test", () => {
  var explorerViewModel = new ExplorerViewModel(dbDescription, {});
  explorerViewModel.getDataCallback = (entityId, attributes, limit, offset, ready) => {
    if (limit > 1) {
      ready([
        { table_key: 1, f1: "one", f2: "first" },
        { table_key: 2, f1: "two", f2: "second" }
      ]);
    }
    if (limit == 1) {
      ready([
        { table_key: 1, f1: "one", f3: "third" },
      ]);
    }
  };
  const { container } = render(<Explorer model={explorerViewModel} />);
  act(() => explorerViewModel.start("table"));
  expect(container.firstChild).toMatchSnapshot();
  fireEvent.click(container.querySelector("tbody tr"));
  expect(container.firstChild).toMatchSnapshot();
});
