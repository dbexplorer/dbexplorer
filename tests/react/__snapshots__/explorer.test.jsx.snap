// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explorer test 1`] = `<div />`;

exports[`Form panel test 1`] = `
<div>
  <div>
    Close
  </div>
  <div>
    <form
      action="#"
    />
  </div>
</div>
`;

exports[`Table panel test 1`] = `
<div>
  <div>
    Close
  </div>
  <div>
    <table>
      <thead>
        <tr />
      </thead>
      <tbody />
      <tfoot>
        <button>
          Load more data...
        </button>
      </tfoot>
    </table>
  </div>
</div>
`;
